kite-tools-parent/kite-tools/src/main/java/org/kitesdk/cli/commands/TarImportCommand.java

```diff
@@ -0,0 +1,167 @@
+    Preconditions.checkArgument(targets != null && targets.size() == 2,
+        "Tar path and target dataset name are required.");
+
+    Path source = qualifiedPath(targets.get(0));
+    FileSystem sourceFS = source.getFileSystem(getConf());
+    Preconditions.checkArgument(sourceFS.exists(source),
+        "Tar path does not exist: " + source);
+
+    Preconditions.checkArgument(SUPPORTED_TAR_COMPRESSION_TYPES.contains(compressionType),
+        "Compression type " + compressionType + " is not supported");
+
+    String datasetName = targets.get(1);
+
+    int success = 0;
+
+    DatasetRepository repo = getDatasetRepository();
+
+    Dataset<GenericRecord> tarImport = null;
+    if (!repo.exists(namespace, datasetName)) {
+      console.info("Creating dataset {} in repo {}", datasetName, repo.getUri());
+      tarImport = repo.create(namespace, datasetName,
+          new DatasetDescriptor.Builder()
+              .schema(new Schema.Parser().parse(
+                  "{\n" +
+                      "  \"type\": \"record\",\n" +
+                      "  \"name\": \"tarimport\",\n" +
+                      "  \"fields\": [\n" +
+                      "  {\n" +
+                      "    \"type\": \"string\",\n" +
+                      "    \"name\": \"filename\"\n" +
+                      "  },\n" +
+                      "  {\n" +
+                      "    \"type\": \"bytes\",\n" +
+                      "    \"name\": \"filecontent\"\n" +
+                      "  }\n" +
+                      "  ]\n" +
+                      "}"
+              ))
+              .build()
+      );
+    } else {
+      console.info("Using existing dataset {} at {}", datasetName, repo.getUri());
+      tarImport = repo.load(namespace, datasetName);
+    }
+
+    DatasetWriter<GenericRecord> writer = tarImport.newWriter();
+
+    // Create a Tar input stream wrapped in appropriate decompressor
+    // Enhancement would be to use native compression libs
+    TarInputStream tis = null;
+    if (compressionType.equals("gzip")) {
+      tis = new TarInputStream(new GzipCompressorInputStream(sourceFS.open(source)));
+    } else if (compressionType.equals("bzip2")) {
+      tis = new TarInputStream(new BZip2CompressorInputStream(sourceFS.open(source)));
+    } else {
+      tis = new TarInputStream(sourceFS.open(source));
+    }
+    TarEntry entry = null;
+    GenericRecordBuilder builder = new GenericRecordBuilder(tarImport.getDescriptor().getSchema());
+
+    try {
+      int count = 0;
+      boolean readError = false;
+      while ((entry = tis.getNextEntry()) != null && !readError) {
+        if (!entry.isDirectory()) {
+          long size = entry.getSize();
+          byte[] buf = new byte[(int) size];
+          int read = 0;
+          int br = 0;
+          // TarInputStream does not always read what you ask for so loop until it does
+          // Note that this requires the entire tar entry (not entire file) to fit into memory
```

Comment:
Avro records are not split across blocks, but blocks have a minimum size, not a fixed size. So things will work fine even if you have very large records - it's just that splitting is less efficient when reading (as Ryan notes). As the expected use case is importing tar files from local file systems, I don't think we need to do anything particularly complicated here. Maybe just warn if a record is bigger than an HDFS block?

As for streaming the bytes - there's no mechanism for doing that since you have to build the GenericRecord in memory. I could imagine a specialized version of GenericRecord that uses an InputStream instead of a byte array for an Avro bytes type, but again I don't think we need to do this given the expected use case (where individual records are not greater than an HDFS block).

Also, you might use IOUtils#readFully from Commons Compress that will read a given number of bytes into a byte array: https://commons.apache.org/proper/commons-compress/javadocs/api-1.9/index.html?org/apache/commons/compress/utils/IOUtils.html